internal/api/client.go

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+package api
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"math/rand"
+	"net/http"
+	neturl "net/url"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/microsoft/terraform-provider-power-platform/internal/config"
+	"github.com/microsoft/terraform-provider-power-platform/internal/customerrors"
+	"github.com/microsoft/terraform-provider-power-platform/internal/helpers"
+	"github.com/microsoft/terraform-provider-power-platform/internal/helpers/array"
+)
+
+// ProviderClient is a wrapper around the API client that provides additional helper methods.
+type ProviderClient struct {
+	Config *config.ProviderConfig
+	Api    *Client
+}
+
+// GetConfig returns the provider configuration.
+func (client *Client) GetConfig() *config.ProviderConfig {
+	return client.Config
+}
+
+// Client is a base client for specific API clients implmented in services.
+type Client struct {
+	Config   *config.ProviderConfig
+	BaseAuth *Auth
+}
+
+// ApiHttpResponse is a wrapper around http.Response that provides additional helper methods.
+func NewApiClientBase(providerConfig *config.ProviderConfig, baseAuth *Auth) *Client {
+	return &Client{
+		Config:   providerConfig,
+		BaseAuth: baseAuth,
+	}
+}
+
+var retryableStatusCodes = []int{
+	http.StatusUnauthorized,    // 401 is retryable because the token may have expired.
+	http.StatusRequestTimeout,  // 408 is retryable because the request may have timed out.
+	http.StatusTooEarly,        // 425 is retryable because the request may have been rate limited.
+	http.StatusTooManyRequests, // 429 is retryable because the request may have been rate limited.
+	http.StatusInternalServerError,
```

Comment:
and the others we don't comment :)?

Reply:
Thought the rest of the code was "self commenting" 😂.  Added some more comments